### PR TITLE
Server logging (close #2190)

### LINKED ISSUE #2190: onResponse not called for AJAX requests when hook constructor receives responseEventConfigureOpts
### What is your Test Scenario?
I implemented a custom hook extending `RequestHook`. In the hook's constructor I call `super(null, { includeHeaders: true })` to listen to all requests and use the response headers.

### What is the Current behavior?
The `onResponse` method is only invoked for non-ajax requests.
 
### What is the Expected behavior?
The `onResponse` method is invoked for all requests.
 
### What is your web application and your TestCafe test code?
I created a repo with the sample test: https://github.com/fedebertolini/testcafe-hooks-issue

My hook receives the `responseEventConfigureOpts` config as parameter, and will call `super` passing this object. I created three simple tests using the hook in 3 different ways:

1. passing `undefined`
2. passing an empty object
3. passing `{ includeHeaders: true}`

In `onResponse` I log the url of the ajax requests only. Only in the first case I can see the requests being loaded:

Screenshot 2019-11-24 at 10 37 33

<details>
<summary>Your complete test code (or attach your test files):</summary>

 
```js
 import { RequestHook } from 'testcafe';

export default class CustomHook extends RequestHook {
  constructor(config) {
    super(null, config);
    this.networkRequests = {};
  }

  onRequest(event) {
    const id = event._requestContext.requestId;
    this.networkRequests[id] = {
      url: event._requestInfo.url,
      isAjax: event.isAjax,
    };
    if (event.isAjax) {
        console.log('request: ' + event._requestInfo.url);
    }
  }

  onResponse(event) {
    const request = this.networkRequests[event.requestId];
    if (!request) return;

    if (request.isAjax) {
        console.log('response: ' + request.url);
    }

    delete this.networkRequests[event.requestId];
  }
}
```

```js
import customHook from './customHook';

fixture('google').page('https://www.google.com/');

test('With out config', () => {}).requestHooks(new customHook());

test('With empty config', () => {}).requestHooks(new customHook({}));

test('With includeHeaders', () => {}).requestHooks(new customHook({ includeHeaders: true}));

```
</details>

<details>
<summary>Your complete configuration file (if any):</summary>

```
No config
```
</details>

<details>
<summary>Your complete test report:</summary>

```
> testcafe chrome index.js

 Running tests in:
 - Chrome 78.0.3904.108 / macOS 10.14.6

 google
request: https://www.google.de/domainless/read?igu=1
response: https://www.google.de/domainless/read?igu=1
request: https://www.google.com/domainless/write?igu=1&data=&xsrf=AO61pXTlD4i518O7D3bSIKfaIcEDFzoNhA:1574588204976
response: https://www.google.com/domainless/write?igu=1&data=&xsrf=AO61pXTlD4i518O7D3bSIKfaIcEDFzoNhA:1574588204976
 ✓ With out config
request: https://www.google.de/domainless/read?igu=1
request: https://www.google.com/domainless/write?igu=1&data=&xsrf=AO61pXS3sSSX7Afhd3tGK3Y9R5ICC57Nmg:1574588208441
 ✓ With empty config
request: https://www.google.de/domainless/read?igu=1
request: https://www.google.com/domainless/write?igu=1&data=&xsrf=AO61pXRFp4rmwJ9BrdfVqefslQilGyyQ8w:1574588209606
 ✓ With includeHeaders

 3 passed (6s)
```
</details>

### Steps to Reproduce:

1. Clone https://github.com/fedebertolini/testcafe-hooks-issue
2. Run `npm i`
3. Run `npm start` (assuming you have chrome installed)
4. Check the logs
 
### Your Environment details:
 
* testcafe version:                   `1.7.0`
* node.js version:                    `v12.13.0`
* command-line arguments:  `testcafe chrome index.js`
* browser name and version: `chrome 78.0.3904.108`
* platform and version:          `macOS 10.14.6`

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

**Comment by LavrovArtem** [comment]
Debug categories:
* `hammerhead:proxy`
* `hammerhead:destination`
* `hammerhead:service-message`

File: src/request-pipeline/context.ts [hunk 1]
```diff
@@ -381,11 +384,16 @@ export default class RequestPipelineContext {
...
 
+        logger.proxy('Proxy response %s %d %j', this.requestId, this.destRes.statusCode, headers);
+
         res.writeHead(this.destRes.statusCode as number, headers);
         res.addTrailers(this.destRes.trailers as http.OutgoingHttpHeaders);
     }
 
     async mockResponse (): Promise<void> {
+        // @ts-ignore
+        logger.destination('Destination request is mocking %s %s %j', this.requestId, this.mock.statusCode, this.mock.headers);
```
Comment (miherlosev) [inline src/request-pipeline/context.ts:395]: Make the `ResponseMock.statusCode` and `ResponseMock.headers` properties public and remove `ts-ignore` comment.

File: src/utils/logger.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    if (contentInfo.charset)
+        stringifiedInfoArr.push('charset: ' + contentInfo.charset.get());
+
+    stringifiedInfoArr.push('encoding: ' + contentInfo.encoding);
+    stringifiedInfoArr.push('requireProcessing: ' + contentInfo.requireProcessing);
+
+    return `{ ${stringifiedInfoArr.join(', ')} }`;
+};
+
+const hammerhead        = debug('hammerhead');
```
Comment (miherlosev) [inline src/utils/logger.ts:50]: Remove extra spaces.

File: src/request-pipeline/context.ts [hunk 3]
```diff
@@ -381,11 +384,15 @@ export default class RequestPipelineContext {
...
             headerTransforms.setupPreventCachingHeaders(headers);
 
+        logger.proxy('Proxy response %s %d %j', this.requestId, this.destRes.statusCode, headers);
+
         res.writeHead(this.destRes.statusCode as number, headers);
         res.addTrailers(this.destRes.trailers as http.OutgoingHttpHeaders);
     }
 
     async mockResponse (): Promise<void> {
+        logger.destination('Destination request is mocking %s %s %j', this.requestId, this.mock.statusCode, this.mock.headers);
```
Comment (VasilyStrelyaev) [inline src/request-pipeline/context.ts:394]: `Destination request is mocked` probably?

File: src/request-pipeline/destination-request/index.ts [hunk 4]
```diff
@@ -105,24 +109,32 @@ export default class DestinationRequest extends EventEmitter implements Destinat
...
+        logger.destination('Destination upgrade %s %d %j', this.opts.requestId, res.statusCode, res.headers);
+
         if (head && head.length)
             socket.unshift(head);
 
         this._onResponse(res);
     }
 
     async _resendWithCredentials (res): Promise<void> {
+        logger.destination('Destination request resend request with credentials %s', this.opts.requestId);
```
Comment (VasilyStrelyaev) [inline src/request-pipeline/destination-request/index.ts:136]: `Destination request resent with credentials`?

File: src/request-pipeline/index.ts [hunk 5]
```diff
@@ -93,6 +96,7 @@ const stages = [
             await ctx.session.callRequestEventCallback(RequestEventNames.onConfigureResponse, rule, configureResponseEvent);
             await callOnResponseEventCallbackForFailedSameOriginCheck(ctx, rule, ConfigureResponseEventOptions.DEFAULT);
         });
+        logger.proxy('Proxy CORS check failed %s, respond 222', ctx.requestId);
```
Comment (VasilyStrelyaev) [inline src/request-pipeline/index.ts:99]: `Proxy CORS check failed %s, response 222` - if we got 222
`Proxy CORS check failed %s, responding 222` - if we are going to send 222

File: src/request-pipeline/index.ts [hunk 6]
```diff
@@ -192,7 +196,10 @@ const stages = [
 export async function run (req: http.IncomingMessage, res: http.ServerResponse | net.Socket, serverInfo: ServerInfo, openSessions: Map<string, Session>): Promise<void> {
     const ctx = new RequestPipelineContext(req, res, serverInfo);
 
+    logger.proxy('Proxy request %s %s %s %j', ctx.requestId, ctx.req.method, ctx.req.url, ctx.req.headers);
+
     if (!ctx.dispatch(openSessions)) {
+        logger.proxy('Proxy error: request to proxy cannot be dispatched %s, respond 404', ctx.requestId);
```
Comment (VasilyStrelyaev) [inline src/request-pipeline/index.ts:202]: same as above

File: src/request-pipeline/utils.ts [hunk 7]
```diff
@@ -47,6 +48,9 @@ export function sendRequest (ctx: RequestPipelineContext) {
         });
 
         req.on('fatalError', err => {
+            if (ctx.isFileProtocol)
+                logger.destination('File reading error %s %o', ctx.requestId, err);
```
Comment (VasilyStrelyaev) [inline src/request-pipeline/utils.ts:52]: `File read error`

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Pull request merged**